### PR TITLE
migrate reentrancy to new directory structure

### PR DESCRIPTION
**Pull request by jtriley-eth**
## Changes

The following changes have been made:

- Reentrancy library has been migrated from the std library to sway-libs
- Reentrancy tests have been migrated
- A new directory, `test_artifacts` has been introduced for multi-contract tests
- The build script (`.github/scripts/build.sh`) has been modified to build `test_artifacts`

## Notes

- It may be worth reviewing the rest of the ci pipeline to ensure the new directory is handled properly

## Related Issues

Part 1 of [Issue 3217 in Fuellabs/sway](https://github.com/FuelLabs/sway/issues/3217)

### CODE REVIEW

**Comment by nfurfaro** [comment]
@jtriley-eth Do you want to perhaps improve the docs for `reentrancy_guard` while you're touching these files? 
I opened [this PR](https://github.com/FuelLabs/sway/pull/3893) to do that, but it makes more sense to make the changes here.

File: libs/reentrancy/SPECIFICATION.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Overview
+
+This document provides an overview of the Reentrancy library.
+
+It outlines the use cases and specification.
+
+## Use Cases
+
+The reentrancy check is used to check if a contract ID has been called more than
+once.
```
Comment (nfurfaro) [inline libs/reentrancy/SPECIFICATION.md:10]:
```suggestion
once in the current call stack.
```

File: libs/reentrancy/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+## Using the Library
+
+### Using the Reentrancy Guard
+
+Once imported, using the Reentrancy Library can be done by calling one of the two functions. For
+more information, see the [specification](./SPECIFICATION.md).
+
+- `is_reentrant() -> bool`
+- `reentrancy_guard()`
```
Comment (bitzoic) [inline libs/reentrancy/README.md:17]: How do you feel about changing this to an example like so?

```rust
fn my_function() {
     reentrancy_guard();
     // My code here
}
```

File: tests/src/reentrancy/mod.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            .await
+            .unwrap();
+    }
+}
+
+mod revert {
+    use super::*;
+
+    #[tokio::test]
+    #[should_panic]
```
Comment (bitzoic) [inline tests/src/reentrancy/mod.rs:91]: Are you able to add the expected panic message?

`#[should_panic(expected = "NonReentrant")]`

File: libs/reentrancy/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Overview
+
+The Reentrancy library provides an API to check for and disallow reentrancy on a contract.
+
+More information can be found in the [specification](./SPECIFICATION.md).
+
+## Known Issues
```
Comment (bitzoic) [inline libs/reentrancy/README.md:7]:
```suggestion

```

File: libs/reentrancy/src/lib.sw [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+use std::call_frames::*;
+use std::registers::frame_ptr;
+
+pub enum ReentrancyError {
+    NonReentrant: (),
+}
+/// Reverts if the reentrancy pattern is detected in the contract in which this is called.
+/// Not needed if the Checks-Effects-Interactions (CEI) pattern is followed (as prompted by the
+/// compiler).
+/// > Caution: While this can protect against both single-function reentrancy and cross-function
```
Comment (bitzoic) [inline libs/reentrancy/src/lib.sw:15]: I feel like this warning should be included in the README or SPECIFICATION

**Review by bitzoic — COMMENTED** [review]
Could you add the library to the [source README](https://github.com/FuelLabs/sway-libs/blob/master/README.md) so users are able to easily find it?

Besides a few minor comments LGTM! 🚀

**Review by nfurfaro — COMMENTED** [review]
One minor nit, otherwise looks good!

File: libs/reentrancy/src/lib.sw (hunk 5 above)
Comment (jtriley-eth) [reply libs/reentrancy/src/lib.sw:15]: Good catch, added to Known Issues.

File: libs/reentrancy/README.md (hunk 2 above)
Comment (jtriley-eth) [reply libs/reentrancy/README.md:20]: Good idea, done!

File: tests/src/reentrancy/mod.rs (hunk 3 above)
Comment (jtriley-eth) [reply tests/src/reentrancy/mod.rs:91]: There is some issue with bubbling up this error message. The call stack is as follows:

- `attacker.launch_thwarted_attack_1` -> `target.intra_contract_call`
- `target.intra_contract_call` -> `target.cross_function_reentrance_denied`
- `target.cross_function_reentrance_denied` -> `attacker.evil_callback_3`
- `attacker.evil_callback_3` -> `attacker.launch_thwarted_attack_1`
- `attacker.launch_thwarted_attack_1` -> `target.intra_contract_call`
- `target.intra_contract_call` -> `target.reentrance_denied` (throws "NonReentrant")

But using `#[should_panic(expected = "NonReentrant")]` results in `expected substring: "NonReentrant"`

Comment (bitzoic) [reply tests/src/reentrancy/mod.rs:91]: This likely has to do with https://github.com/FuelLabs/fuels-rs/pull/762. 

The repo will need to be bumped to update this.

**Review by bitzoic — APPROVED** [review]
LGTM!

**Pull request merged**